crates/apollo-router/src/apollo_router.rs

```diff
@@ -0,0 +1,125 @@
+use apollo_router_core::prelude::graphql::*;
+use derivative::Derivative;
+use futures::prelude::*;
+use std::sync::Arc;
+use tracing::Instrument;
+use tracing_futures::WithSubscriber;
+
+/// The default router of Apollo, suitable for most use cases.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct ApolloRouter {
+    #[derivative(Debug = "ignore")]
+    naive_introspection: NaiveIntrospection,
+    query_planner: Arc<dyn QueryPlanner>,
+    service_registry: Arc<dyn ServiceRegistry>,
+    schema: Arc<Schema>,
+}
+
+impl ApolloRouter {
+    /// Create an [`ApolloRouter`] instance used to execute a GraphQL query.
+    pub fn new(
+        query_planner: Arc<dyn QueryPlanner>,
+        service_registry: Arc<dyn ServiceRegistry>,
+        schema: Arc<Schema>,
+    ) -> Self {
+        Self {
+            naive_introspection: NaiveIntrospection::from_schema(&schema),
+            query_planner,
+            service_registry,
+            schema,
+        }
+    }
+}
+
+impl Router<ApolloRoute> for ApolloRouter {
+    fn create_route(
```

Comment:
not sure about calling it `create_route` and return an `ApolloRoute`. Can we have a word around "query" instead? Since this is handling one query

Reply:
It's not "handling" the whole query. It's creating a `Route` instance which will "handle" (execute) the query.

Reply:
Naming is hard but I like the `Router`'s `create_route()` literally returning a newly created `Route`. If others have suggestions I will change to whatever the consensus go.

Reply:
maybe an Execution or something like that? In databases what comes after the query planner is the executor, in charge of executing the steps.

Reply:
I talked to @o0Ignition0o and I realized that the naming of Router and Route wasn't as obvious to me as it is to others :sweat: I'll try to explain the reasoning better:

> [...] graph router for a federated graph.

Apollo Router is a "graph router". What does a router do? It routes things. So the core of Apollo Router is to be a "router" (for graphs).

So what does a router do? It creates a path of execution, aka "route". The router create a specific "route" for a specific request which can then be executed.

> Rename FederatedGraph to Router

The whole point of the PR is to split the process of creation of this route from the execution of this route.

FederatedGraph => takes a request, query the services, return a response stream
Router => takes a request, make a route
Route => can be executed to get the response stream

Reply:
(So the function on Router is called create_route because it creates a new route. While the function Route is called execute because it actually execute the process)

Reply:
I think the truth is the router isn't a router, It's a sequencer or an orchestrator or something like that. What it means is we're not actually routing queries, but Building an execution plan, and running it. This is IMO the reason why I was confused with ApolloRouter and ApolloRoute, also the apollo prefix doesn't make much sense here imo.

now it looks like an ApolloRoute executes a plan against uh services, and it passes _a lot_ of things to it as parameters. I think there's room for refactoring, but I don't know how exactly. I'll try to let you know if I can think of anything. The short version is that imo query_plan.execute() has too big parameters, so we might be able to move things around and come up with a more elegant set of functions

Reply:
> the apollo prefix doesn't make much sense here imo.

The Apollo prefix is to differentiate the core Router trait from the struct provided by Apollo (the company) for general purpose.

> it looks like an ApolloRoute [...] it passes a lot of things to it as parameters.

A lot of things are passed in order to ensure the object will have (shared) ownership of the resources it requires to execute the process: the query plan itself, the service registry, the schema and a cached parsed query.

The request object though should be provided in argument as @BrynCooke mention as this might holds variables specific to a particular request.

> ApolloRoute executes a plan against uh services

Not services. A request. But that will be more obvious when I move request to the function arguments.

> The short version is that imo query_plan.execute() has too big parameters, so we might be able to move things around and come up with a more elegant set of functions

These are all immutable Arcs (low in memory usage) and meant to be used in scope of the request execution. Everything you need to be able to execute the query should be there. It's also important that these are internals to the Route itself as it can be customized differently by a different kind of Route.

Of course it's possible to make things simple: remove the lib, keep only the bin, remove all the traits. But my goal is to make this extensible to the user. Traits are a good way to make things extensible.

Reply:
It's still very confusing for me. Router is an overloaded word, so we try to match it with a common meaning (the first one for me is a network component). The project's name is Apollo router, and it defines a HTTP server, available as library or executable, that receives multiple graphql queries over connections, and executes them. This a struct also called ApolloRouter, that does not receives HTTP requests directly, but we can pass it a request and it can return a future of Response, but actually it's not a HTTP request handler or [service](https://docs.rs/tower-service/0.3.1/tower_service/trait.Service.html), because it can return a route?

> It creates a path of execution, aka "route"

That is not evident, unfortunately.

I feel the issue is that we're confusing internal implementation details with user facing API.
If I'm a user that wants to build a custom router, with my own network layer (example: one endpoint receives graphql, others my own app). I need a [Service](https://docs.rs/hyper/0.14.15/hyper/service/trait.Service.html) I can call if I see graphql requests, maybe add middlewares. I give that service a request and it handles executing and returning a response, be it an introspection query or a larger query.

If I'm a user that wants to build a custom service, maybe to modify the query plan, filter info from introspection, then I need a structure that can:
- given a query, give me an introspection response
- given a query, give me a query plan
- given a query plan, executes the plan and gives me the response

That structure would let me handle the response afterwards. For that, the words "query planner", "orchestrator", "executor" work better, I think

Reply:
> The Apollo prefix is to differentiate the core Router trait from the struct provided by Apollo (the company) for general purpose.

The Apollo router is the Router :wink:

Reply:
> The Apollo router is the Router :wink: 

And The Router answers to all queries...... :wink:

Reply:
> The project's name is Apollo router, and it defines a HTTP server, available as library or executable, that receives multiple graphql queries over connections, and executes them. 

According to the README:
> The **Apollo Router** is a configurable, high-performance **graph router** for a [federated graph](https://www.apollographql.com/docs/federation/).

So it's actually a **graph** router. 

> This a struct also called ApolloRouter, that does not receives HTTP requests directly, but we can pass it a request and it can return a future of Response

Yes!! Almost. You pass it a `graphql::Request` (not http request) and it returns a `graphql::Route` in case of success or a `graphql::ResponseStream`.

> but actually it's not a HTTP request handler or service, because it can return a route?

Exactly. It is agnostic of the HTTP layer. It's above http, it's a GraphQL thingy. It doesn't return any route, it returns a graphql::Route.

> > It creates a path of execution, aka "route"
>
> That is not evident, unfortunately.

imo: a router does routing.

tr.v. rout·ed, rout·ing, routes
1. To send or forward by a specific route. See Synonyms at send.
2. To schedule the order of (a sequence of procedures).

@o0Ignition0o: this is starting to sounds a lot like "It's a sequencer or an orchestrator or something like that.")
@Geal: "the words "query planner", "orchestrator", "executor" work better, I think"

I don't know I'm not a native English speaker but a router doing routing by creating routes that can be executed still sounds good to me.

> I feel the issue is that we're confusing internal implementation details with user facing API.

I made this PR to improve the user facing API. They can now implement their own implementation of Router or Route. This was not possible before. They can make an `ApolloRouter<MyCustomRoute>` or a `MyCustomRouter<ApolloRoute>`. So they can hook right before the query execution itself or right before the route preparation itself.

(API might not be complete for `ApolloRouter<MyCustomRoute>` but that's the general idea)

Reply:
It's starting to feel bikesheddy, and that should not block this PR, especially if we want to modify those parts of the code and avoid conflicts.
Let's go with those words for now. Can you open an issue and tag people to discuss how we should name that part and revisit that later?

Reply:
Ok i didn't know of meaning number 2., it makes sense!

I'm all for merging this as well, Its much better than it was before, and if I understand it correctly you meant this PR as a first step towards a more general refactoring.

+1 for me as well

Reply:
Please tag me into that convo!  Happy to discuss. 😉